[gui] Extend settings dialog functions to show seperator with group name

AlwinEsch opened the pull request:
With this change it become possible to add a name field on the begin of a new group and with/without help description on other place.

Also is <code>virtual std::string GetSettingsLabel(CSetting *pSetting);</code> added which can be used as virtual in dialog class where the base ask about the string, there is it possible to return a own string direct which is not related to the string id's.

Further is the function <code>AddInfoStringButton</code> added which use a button field as field for infos.

Is taken from Audio DSP #4402.

Montellese commented on xbmc/settings/dialogs/GUIDialogSettingsBase.cpp:
```diff
@@ -539,6 +552,11 @@ std::set<std::string> CGUIDialogSettingsBase::CreateSettings()
   return settingMap;
 }
 
+string CGUIDialogSettingsBase::GetSettingsLabel(CSetting *pSetting)
```
What exactly is the use case here? It's just an indirection and doesn't seem to be used anywhere else.

AlwinEsch replied:
Is part of the DSP, is required to have the possibility to give a string direct instead the id, it can be added in the dialog and a string becomes returned for this label.
Is needed there to show addon strings.

Montellese commented on xbmc/settings/dialogs/GUIDialogSettingsManualBase.cpp:
```diff
@@ -239,6 +245,24 @@ CSettingAction* CGUIDialogSettingsManualBase::AddButton(CSettingGroup *group, co
   return setting;
 }
 
+CSettingString* CGUIDialogSettingsManualBase::AddInfoStringButton(CSettingGroup *group, const std::string &id, int label, int level, std::string info, bool delayed /* = false */,
```
Does `delayed` make any sense here? Isn't it always a static string?

AlwinEsch replied:
Thought if it is possible to add, the <code>delayed</code>, but becomes currently in DSP not used.

The function is used to show the cpu usage for DSP process and if present to open addon dialog.

Montellese replied:
The idea of `delayed` is that it takes a bit before the changed value of a setting takes effect. This makes sense for edit controls where the user can type or for spinners where the user might want to spin through all the available options without immediately seeing the effect. I don't see how that makes any sense for an  info label which can't be changed by the user.

Montellese commented on xbmc/settings/dialogs/GUIDialogSettingsManualBase.cpp:
```diff
@@ -89,7 +89,7 @@ CSettingCategory* CGUIDialogSettingsManualBase::AddCategory(const std::string &i
 }
 
-CSettingGroup* CGUIDialogSettingsManualBase::AddGroup(CSettingCategory *category)
+CSettingGroup* CGUIDialogSettingsManualBase::AddGroup(CSettingCategory *category, int label, int help, bool hideSeparator)
```
Please fill in the default values of the parameters.

Montellese commented on xbmc/settings/lib/SettingSection.h:
```diff
@@ -69,6 +121,10 @@ class CSettingGroup : public ISetting
   void AddSettings(const SettingList &settings);
 
 private:
+  int m_label;
```
Maybe it's time to move `m_label`, `m_help` and all their getters and setters into `ISetting` since it's now used by all of its subclasses (`CSettingSection`, `CSettingCategory`, `CSettingGroup` and `CSetting`).

Montellese commented on xbmc/settings/lib/SettingSection.cpp:
```diff
@@ -78,6 +81,20 @@ bool CSettingGroup::Deserialize(const TiXmlNode *node, bool update /* = false */
+    m_label = tmp;
+  if (element->QueryIntAttribute(SETTING_XML_ATTR_HELP, &tmp) == TIXML_SUCCESS && tmp > 0)
+    m_help = tmp;
+  if (element->QueryIntAttribute(SETTING_XML_ATTR_BEFORE, &tmp) == TIXML_SUCCESS && tmp > 0)
```
You can't abuse `SETTING_XML_ATTR_BEFORE` here since it's already being used to be able to insert a group into a specific place.

AlwinEsch replied:
Thanks for the feedback. I'm currently doing a rework. Good to have it separated from DSP :-) .
Is <code>#define SETTING_XML_ATTR_HIDE_SEPARATOR "hideseparator"</code> OK for you and not to long.
Sorry for few bad parts on it, have not recorded all the changes.

Montellese commented on xbmc/settings/lib/SettingDefinitions.h:
```diff
@@ -64,6 +64,7 @@
 #define SETTING_XML_ATTR_SETTING      "setting"
 #define SETTING_XML_ATTR_BEFORE       "before"
 #define SETTING_XML_ATTR_AFTER        "after"
+#define SETTING_XML_ATTR_HIDE         "hide"
```
This isn't a very obvious name as it could be interpreted as "hide the whole group".
Also I'm not sure why you need two options. What different cases do you want to cover?

AlwinEsch replied:
Becomes changed, is thought to hide separator and is a bit to short from name to see this.

Montellese commented:
I'm wondering if the new `hideseparator` and `beforeseparator` options for setting groups are really necessary. IMO we should try to make all settings dialogs look as similar as possible and such options will simply lead to decisions made on personal taste which will leave us with one dialog using one style and another dialog using the other style.

AlwinEsch commented:
Here on the screenshot from DSP you see a place of usage for it, there in the process chain are groups which not need a separator and on others is the string is placed behind.

https://cloud.githubusercontent.com/assets/6879739/2571813/d2d6db0e-b90a-11e3-9a8d-bf928e26956a.png

Want to have also the possibility for bigger setting dialogs (maybe on a addon) to make a group name field on the top and in the center of the group and was no big way to allow also for settings.xml.

Montellese commented:
It looks to me like you are trying to do with groups (and subgroups) what you could do with categories. IMO having multiple short lists (per category) with settings is much easier to use than one long list where you have to scroll a lot to see all the available options.

Furthermore from that screenshot I only see the usage of the `hideseparator` option and still don't see why it needs the `beforeseparator` option. You don't even provide a way to set that through `CGUIDialogSettingsManualBase::AddGroup()` and since nobody is using it through `settings.xml` I don't see how it's needed.

Montellese commented on xbmc/settings/windows/GUIControlSettings.cpp:
```diff
@@ -543,7 +544,7 @@ bool CGUIControlButtonSetting::OnClick()
     else if (controlFormat == "path")
       SetValid(GetPath((CSettingPath *)m_pSetting));
-    else if (controlFormat == "action")
+    else if (controlFormat == "action" || controlFormat == "infoaction")
```
Will you make use of the `OnSettingAction()` callback for an `infoaction` setting or is it just meant to be a label/text?

AlwinEsch replied:
Is handled exactly the same as a normal button with action only, but on line 618 the label becomes continuously updated with it. Don't want to do this for everything, thats why separate name.

Montellese replied:
That doesn't answer my question. Do you have a usecase where you add a setting with the "infoaction" control format and register for the `OnSettingAction()` callback which is triggered when the user activates the label representing the setting? If not it would IMO be better to add a "label" control format with its own `CGUIControlLabelSetting` implementation (from which `CGUIControlButtonSetting` could be derived).

AlwinEsch replied:
Hope to understand you right, you mean is only used as label. Yes and no, is used to open information dialogs on addon (if present). Is a normal button to do somethings, only difference is that the label becomes continuously updated with the cpu usage of the addon dsp mode as string or if not present comes a from KaiToast a information with "Not present from addon...". But is also used on few points only to show information (The first text field on dialog).

AlwinEsch commented:
Thought to have the possibility to add something like this, there and on addon dialogs, if you wish I remove the part for xml.
![bildschirmfoto1](https://cloud.githubusercontent.com/assets/6879739/6016213/f3aadda0-ab7e-11e4-9cb3-65446535d305.png)

Montellese commented:
My point is that IMO it doesn't make sense to have two ways to present named groups. Personally I like the approach in the second screenshot much better (whether left or center aligned) than the first but it all comes down to taste and I want to avoid that some people do it like in your first screenshot and others like in your second.
@ronie / @BigNoid / @da-anda: what's your opinion on this?

da-anda commented:
I like the possibility to add a group label like in the second screenshot - reminds me of &lt;fieldset&gt;> tags in HTML which are also used to group elements and can have an optional label. Good for usability and accessibility as it gives the settings another level of metadata and might even allow to cleanup/simplify some veeeeery long settings labels once the context is indicated by the group label.

As for the "hideseparator" - I don't like it when things influence how stuff is rendered when it shouldn't have control over this. If we need different ways to render groups/categories then we should give them a semantical flag so that skins can decide how this semantical meaning should look like. I did this very same approach on a customer project which should act as an abstract content repository. They needed a way to group content sections in different ways with the aim to show them as tabs, accordeon or whatever when shown on a website. But if the content is rendered in tabs or an accordeon heavily depends on the output media and usecase, so while websites nowadays show tabs and accordeons it's a whole different story for mobile apps where this "layout" decision might not make sense. So it was an absolute no-go that editors create a "tabs" content element, instead they create a group and flag it with a semantic content type, where several of them could end up in the very same layout on a given output media. We could do the same here I think and only give skins an indication what type of content this group/section contains.

AlwinEsch commented:
With this hidesepartor was added from me to have on the "Audio Information" dialog the cpu usage for the different processing groups without it. Maybe one way can be to add a subgroup which not use a separator. 

This information dialog is not direct used to open something, only to see it and not for settings. Thought there as information table is it ok to have strings below separator. If part stays in, then I swap them over to have above as default.

AlwinEsch commented:
Hello, tried here to set string on place requested from skin. Found a way but not sure if it is acceptabe.

Has leaved it still by own "image" and "label" control and added to label only the check for <code>"< reverse >false or true< /reverse >"</code> (name is only temporary, used present one which come most near).

Only other way I see currently is to add a complete new class which include both.

AlwinEsch commented:
Is updated.

AlwinEsch commented:
Hello all, is complete updated the first notes from you are fixed and handling about group label becomes now done by the new CGUISettingsLabelControl class where a label and image together is possible.

See the last commit how it is handled inside the skin.

Only part which is not possible from skin, is to hide the separator, can only be done from settings.xml or programmed dialog, With the alignment of the image is everything possible what a skin designed want to do.

@Montellese, sorry about my answers to your comments. Sometimes is my english a bit confusing for me, I'm normally aircraft mechanic where english is normal and understand there from technical side everything, but with normal things or other regions (Programming) I'm sometimes a small stupid guy ;-) .

AlwinEsch commented:
Can the CGUISettingsLabelControl class becomes reviewed? I don't to want have not needed calls or mistakes.

Paxxi commented on xbmc/guilib/GUISettingsLabelControl.h:
```diff
@@ -0,0 +1,87 @@
+\brief
+*/
+
+#ifndef GUILIB_GUISETTINGSLABELCONTROL_H
```
you can remove these ifndef/endif, pragma once is enough

Paxxi commented on xbmc/guilib/GUISettingsLabelControl.h:
```diff
@@ -0,0 +1,87 @@
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
```
include < string > as it's used further down

Paxxi commented on xbmc/guilib/GUISettingsLabelControl.cpp:
```diff
@@ -0,0 +1,199 @@
+#include "GUIDialog.h"
+#include "GUIFontManager.h"
+
+using namespace std;
```
Dont't drag in all of std, we use it all over the place but we can avoid old sins in new files :)

Paxxi commented on xbmc/guilib/GUISettingsLabelControl.cpp:
```diff
@@ -0,0 +1,199 @@
+  m_image.SetInvalid();
+}
+
+void CGUISettingsLabelControl::SetLabel(const string &label)
```
prefix with std:: for string

Paxxi commented:
code looks good overall except for the line comments I added. If the design is sound I leave to @Montellese to decide.
It would be nice to see dynamic_cast rather than C-style casts but I can't fault you for following the style already used in the files you modified.

AlwinEsch commented:
Is updated, thanks @Paxxi

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
OSX build is added, thanks @mkortstiege

jenkins build this please again

Montellese commented:
@AlwinEsch:
* Please split the changes to the settings library (`xbmc/settings/lib/`) into two separate commit, one for moving `label` and `help` into `ISetting` and one for the changes to `CSettingGroup`.
* What's the advantage of introducing `CGUISettingsLabelControl`? Wouldn't a simple label control work as well?
* I still don't think that the `delayed` property makes any sense on the `infobutton` control.
* Does `hidevalue` make any sense on the `infobutton` control?
* What exactly is the difference between `button` and `infobutton` apart from the label fallback value?

@ronie: Can you take a look at the Confluence changes? Do you want all the default setting controls to be present in all setting dialogs or do you usually only add the controls that are needed by that specific setting dialog?

AlwinEsch commented:
Hello @Montellese, with CGUISettingsLabelControl is thought to combine label with a separator image in one class and allow on skin to place image in wanted position. 

With delayed was also taken to hold style with all other functions.

The button was taken only from the other, only difference is that right label becomes continues updated and was on begin together. I bring them both together with this as child class.

AlwinEsch commented:
No idea, was done by @mkortstiege on OSX and was moved there to another place, but has done the build test and was OK.

mkortstiege commented:
@Montellese, most likely just a resort on hash. I've already discussed this with @Memphiz the other day already.

ronie commented:
> Can you take a look at the Confluence changes?
the skin changes look ok to me, though i don't understand the need for this infobutton control.

> Do you want all the default setting controls to be present in all setting dialogs or do you usually only add the controls that are needed by that specific setting dialog?
i usually only add what's needed. but that's probably an old habbit from the time when the setting control id's wheren't unified.
these days it might be easier to define a default set of controls, and include them in all settings windows.

Montellese commented:
@ronie: Yeah I don't really see the need for the extra `infobutton` either. It might be ok to have it separate in the code but on the skinning side it should simply use the button control that is already present.

AlwinEsch commented:
Hello @Montellese and @ronie, is updated and added a separate commit for the parts for ISettings, also added in "includes.xml" the "< include name="SettingsList_Controls" >" to use for the dialogs.

AlwinEsch commented:
Is updated again, now ready? :-)

da-anda commented:
I just noticed that this "hideseparator" stuff is still in. Do we want to keep it? I really think such a thing is too skin specific (some skins don't even show a separator and most likely would want to use some other layout changes). So instead of this boolean flag/setting I'd use a contextual tag that describes what kind of settings/content the group contains and then the skin can decide how this type of group should look like (with or without separator, different background, border, ....)

AlwinEsch commented:
This "hideseparator" is not accessible from skin, is only used for internal handling. The other way for me is to add new handling for a normal label to it and removing this.

Gives it not to much contextual tags with this lot of groups in settings? Is even possible but make a so big list of different compare strings not to much for a skin?

Montellese commented:
I still think that we shouldn't let the developer decide whether there should be a separator or not. Letting the skinner decide is out of the scope of this PR and is probably not that easy to achieve but using the separator to create some kind of "super-group" doesn't seem right to me. We already support sections, categories and groups. Adding another level only adds confusion IMO.

da-anda commented:
well "hideseperator" will alter the way things are rendered and as not all skins are using/showing a seperator like Confluence, this internal setting won't have any influence on how these skins show stuff and thus will end up in bad UX. If we can reuse the current separation with sections/categories/groups for this that would ofc be great (like having the possibility to add a label to one of these). If this is not working out well we should find some other way. My suggestion would be a contextual flag that is available to skins with a predefined set of options/values (and if these are just something like "primary", "secondary" and "info").

Btw - such contextual groups would also be nice for mixed search results where we could group by media type. This is ofc out of the scope of this PR - only came to my mind right now.

AlwinEsch commented:
I can maybe use a disabled button for it, you see on the first screenshot what I mean for what it is used. 

Here again
![d2d6db0e-b90a-11e3-9a8d-bf928e26956a](https://cloud.githubusercontent.com/assets/6879739/6186546/01143884-b37b-11e4-9d15-4a77738a061d.png)

To use separator make no sense there for me, becomes to much and without label is more bad.

I understand what you mean some kind of "super-group" is not right, but there is not really a settings field is only information and is really the only point where it make sense. The best is for me to remove this allowance for xml complete.

da-anda commented:
@AlwinEsch in this screenshot it's not really clear to me wh some sections have a separator and some not. If "Process Chain (with CPU usage)" is the parent group for "input:" and "Master processing:" than this it not at all obvious. I don't know the XML structure of this dialog, but IMO "Process Chain..." would have to be a labeled category and "Input:" and "Master processing:" a labeled group (or alike, but at least a nested structure). How this stuff is rendered is up to the skin then. If skins need a way to differentiate between the main Kodi settings window and settings dialogs to adjust rendering, then let's add that instead of the "hideseparator".

AlwinEsch commented:
I currently create a "CSettingVoid* AddLabel(CSettingGroup *group, const std::string &id, int label, int level, bool visible = true);"

This dialog is complete based upon CGUIDialogSettingsManualBase and not related to XML.

Is to split information as "Input stream", "Output stream" and "Workflow"

da-anda commented:
would tabs work like theye are used for many addon settings? Or can we let skins know in which context the settings are rendered (main settings window, dialog/addon settings, ...) so that they could adjust rendering accordingly?

btw - I think this dialog is missing a "close" button or alike. The "x" won't be available with mouse support disabled.

Montellese commented:
@AlwinEsch: I proposed this before: Why don't you use categories for the "super-groups" you are trying to create with the separators? These will be shown as tabs (similar to the addon settings page) and are much more obvious than these groups. You would loose a bit of space at the top of the dialog for the category buttons but your settings list per category would become much smaller and you would probably not even see the scroll bar anymore. Seeing all available settings at one glance is much better anyway.

AlwinEsch commented:
Also on the other Audio or Video Settings is no close button, is the same type as this, but only difference is with sub menus.

This dialog is already in categories
![a3d2bc72-c924-11e3-9f1f-37eddecd9c78](https://cloud.githubusercontent.com/assets/6879739/6187570/3afa4b62-b388-11e4-9637-8283174b9091.png)

I can try to do a complete rewrite but can currently not match directly this but to use the buttons on the left. Problem is also the the amount of buttons is not static.
![dialog-with-categories-x1080](https://cloud.githubusercontent.com/assets/6879739/6187586/6f166d72-b388-11e4-88a1-3ce6aadc0b3a.png)

The flow on this dialog is normally:
Main -----> Content Type
          |---> Listening Mode
          |---> Volume
          |---> Listening Mode setup ------,- Addon Dialog Open button
          |                                                 |-                   "
          |                                                 *- Addon Dialog Open button
          |
          |---> Speaker/Output Setup ----,- Addon Dialog Open button
          |                                                 |-                   "
          |                                                 *- Addon Dialog Open button
          |
          |---> Input stream setup ---------,- Addon Dialog Open button
          |                                                 |-                   "
          |                                                 *- Addon Dialog Open button
          |
          |---> Input Resample setup -----,- Addon Dialog Open button
          |                                                 |-                   "
          |                                                 *- Addon Dialog Open button
          |
          |---> Output resample setup ----,- Addon Dialog Open button
          |                                                 |-                   "
          |                                                 *- Addon Dialog Open button
          |
          |---> Miscellaneous ---------------,- Addon Dialog Open button
          |                                                 |-                   "
          |                                                 *- Addon Dialog Open button
          |
          |---> Audio Information -----------,- Output signal
          |                                                 |- Input signal
          |                                                 |- Process chain
          |                                                 |-- Addon Dialog mode Info button
          |                                                 |--                   "
          |                                                 *-- Addon Dialog mode Info button
          |
          *---> Set as default

The first is categories and the second (right) in groups. Has not seen the possibility for "section" in "CGUIDialogSettingsManualBase"

da-anda commented:
@AlwinEsch for whitespaces see Syntax Highlighting https://guides.github.com/features/mastering-markdown/#GitHub-flavored-markdown

but maybe we should discuss the AudioDSP dialog layouts and structure in the forum to not spam everybody?

About this PR: I'm all in favor of having the possibility to show a group label in settings (not just dialogs but also main Kodi settings page), so maybe we should change this PR to only add the possibility for labels and maybe a group description and do the layout specific stuff in a second step? Would this help you in a way to be able to continue working on open ASDP things?

AlwinEsch commented:
Yes that's was also the main reason for this request not to talk about this dialog. This comes later :-).

AlwinEsch committed:
[gui] Add new CGUISettingsGroupLabelControl class

AlwinEsch committed:
[gui] Extend settings dialog functions to have named groups

AlwinEsch committed:
[settings] Allow coded settings dialogs to define label (needed to pass addon strings)

AlwinEsch committed:
[settings] Add coded settings a button with continuesly updated label

AlwinEsch committed:
[confluence] Add new setting defaults for named groups

AlwinEsch commented:
jenkins build this please
